src/bp/core/services/dialog/instruction/strategy.ts

```diff
@@ -128,14 +136,33 @@ export class ActionStrategy implements InstructionStrategy {
 
     args = _.mapValues(args, value => renderTemplate(value, actionArgs))
 
+    let actionServerId
+    if (chunks.length === 3) {
+      actionServerId = chunks[2]
+    }
+
+    let actionServer: ActionServer | undefined
+    if (actionServerId) {
+      actionServer = await this.actionServersService.getServer(actionServerId)
+      if (!actionServer) {
+        this.logger.warn(`Could not find Action Server with ID: ${actionServerId}`)
+        return ProcessingResult.none()
+      }
+    }
+
     debug.forBot(botId, `[${event.target}] execute action "${actionName}"`)
-    const hasAction = await this.actionService.forBot(botId).hasAction(actionName)
 
     try {
-      if (!hasAction) {
-        throw new Error(`Action "${actionName}" not found, `)
+      if (!actionServerId) {
+        const hasAction = await this.actionService.forBot(botId).hasAction(actionName)
+        if (!hasAction) {
+          throw new Error(`Action "${actionName}" not found, `)
+        }
```

Comment:
I think it would make sense to move part of that logic to the ActionService. Technically, you could just call `this.actionService.hasAction(actionName, serverId?)` then handle the local/remote/etc behind the scene. 

Also, right now you are only checking if the action exists when there's no actionServerId

I still think it should throw an error, failing silently with a log will puzzle some users

Reply:
I won't check if an action exists on a Action Server before starting the execution, as this will introduce slowdowns in flow execution.

If an Action does not exist on the Action Server, the Action Server should return a 404 error.